utils_test: add type check for RemotePackageMgr before is_installed call

YongxueHong opened the pull request:
Add isinstance check to ensure mgr is RemotePackageMgr before calling is_installed() method to prevent potential AttributeError.

ID: 4225

smitterl reviewed (changes requested):
Hi @YongxueHong - There are several things I'd like to address.

1. I fixed the utils_perf.py upstream, it shouldn't be necessary anymore and out of scope for the PR.
2. IIUC, there was an issue introduced in eb091731ebc39efcdcb9e838ee06207844903b84 - Since then the code will skip installing (function `install`) and cleaning up (function `clean`) the stress package in case it's already installed. For that it assumed that the `utils_package.package_manager` had a function `is_installed`. However, `package_manager` can be an instance of `RemotePackageMgr` or `LocalPackageMgr` where the first has the method `is_installed` and the second doesn't.
3. This patch proposes to only check if the package is installed if it's a `RemotePackageMgr` - in other cases the code will just execute `utils_package.package_install` and `utils_package.package_remove` like before the change, e.g. https://github.com/avocado-framework/avocado-vt/blob/065d4ac9442c50ef34531a00df4ad55795601db0/virttest/utils_test/__init__.py

I think if we accept this patch, your commit message should explain 2. and especially 3. for context.

On the other hand, this fix introduces some complexity that shouldn't be necessary. The `is_instance` check bleeds implementation details from `utils_package` into `__init__.py` which shouldn't happen IMO, I was really confused by this during code review and assumed that the other cases where the package manager is not a `RemotePackageMgr` were not handled correctly.

A simpler approach seems to be available: The `LocalPackageMgr` inherits from `avocado.utils.software_manager` and all `software_manager.backends` already implement a `check_installed` method which is actually already used, s. https://github.com/avocado-framework/avocado-vt/blob/27641ba86da6e5ab89a47abfbc3de39c6e60581c/virttest/utils_package.py#L174

As a result, I prefer that you add a method `is_installed` to the `LocalPackageMgr`. This way they have the same interface and functionality. The code becomes simpler works in the same way in both cases, I assume.

Finally, in https://github.com/avocado-framework/avocado-vt/pull/4149 you only handled the stress package but totally omitted the improvement for the dependencies, s. https://github.com/avocado-framework/avocado-vt/blob/27641ba86da6e5ab89a47abfbc3de39c6e60581c/virttest/utils_test/__init__.py#L2511.

Summary: I'd prefer you
# implement `LocalPackageMgr.is_installed` using the existing `software_manager.backends.X.check_installed` function and not do a type check here in `__init__.py`
# also apply this early `is_installed` check to the dependencies

PaulYuuu commented:
Totally I agree @smitterl's idea, adding a alias `def is_installed` in LocalPackageMgr and return `self.check_installed`. But I have another proposal is we can have a new function in utils_package to check if the package is installed, currently utils_package has `package_install`, `package_remove` and `package_upgrade`, having a `is_package_installed` can easy to fix this issue, we can put the logic of using `check_installed` or `is_installed` in the new func.

On the other hand, both `RemotePackageMgr` and `LocalPackageMgr` have `def operate`, they can check the package install status, use them in here should also work, but this means we need to update the operate function by adding `operate_cmd` as argument, and align other arguments.

YongxueHong commented:
> > **Proposed Solutions:**
> > 
> > * Short-Term (Immediate Fix): Since this change introduced a regression, I recommend we revert the commit or implement a minimal workaround to fix the bug quickly and avoid enlarging the scope of impact. CC @PaulYuuu @mcasquer @aliang123 , and I prefer to revert it.
> >   I agree.

Hi @PaulYuuu @mcasquer @aliang123
What are your opinions on reverting the previous commit:https://github.com/avocado-framework/avocado-vt/commit/eb091731ebc39efcdcb9e838ee06207844903b84 ?
Thanks.

> > * Long-Term (Proper Refactor): True consistency would require a significant refactoring effort to establish a common base class or other ways to align the interfaces for both package managers. This is a larger architectural task that should be investigated and planned separately, not as a quick fix. CC @Chwen @dzheng @yicui @nanli @haizhao
> >   Agree, please check my suggestion in the previous reply.
> 
> Thank you!

mcasquer commented:
> What are your opinions on reverting the previous commit:[eb09173](https://github.com/avocado-framework/avocado-vt/commit/eb091731ebc39efcdcb9e838ee06207844903b84) ?

Mmm not sure, I would have to double check the memory stress test cases, especially in Image Mode if the installation logic will be again along with the package check

YongxueHong commented:
Hi @smitterl @dzhengfy @chunfuwen @nanli1 @Yingshun @PaulYuuu @mcasquer @aliang123 
I have implemented a workaround and updated the pull request. Could you please review the changes when you have a moment? Thank you.

Yingshun commented:
@cliping  Can you please help check libvirt migration stress cases with this PR? Thanks!

YongxueHong commented on virttest/utils_test/__init__.py:
```diff
@@ -2508,12 +2508,17 @@ def install(self):
-                )
+            mgr = utils_package.package_manager(self.session, self.dependency_packages)
+            dep_pkgs = [
+                pkg for pkg in self.dependency_packages if not mgr.is_installed(pkg)
```
Hi @PaulYuuu 

Thank you for your feedback. I completely agree with your assessment.

The coupling of the check and install logic within the install method is a long-standing design issue. Unfortunately, due to its many existing call sites, modifying it directly would be a high-risk change that could disrupt numerous test cases. We've previously implemented workarounds for this very reason, as seen in commit [eb091731](https://github.com/avocado-framework/avocado-vt/commit/eb091731ebc39efcdcb9e838ee06207844903b84). CC  @smitterl @dzhengfy @chunfuwen @nanli1 @Yingshun @PaulYuuu @mcasquer @aliang123 @cliping @chloerh

My intention with this PR was to deliver a safe, targeted fix for the regression without altering other public methods and introducing breaking changes. That's why I proposed the initial suggestion to use `isinstance(mgr, utils_package.RemotePackageMgr)` was intended as a quick and immediate workaround to resolve the issue. But there is a new method to check for installation from the @smitterl [comments](https://github.com/avocado-framework/avocado-vt/pull/4208#pullrequestreview-3093626983) , even with the slight redundancy.  However, it looks good for me since my intention is to fix this issue in a short time and with fewer changes for existing call sites.

I believe the root cause is the fragile initial design, which didn't properly decouple functionalities or ensure consistency between `RemotePackageMgr` and `LocalPackageMgr`. Correcting this would require a significant refactoring effort, which is likely beyond the scope of this single PR.

While I believe a full refactor is the right long-term solution, I unfortunately don't have the capacity to take on that larger task at this time. Given that, I see two paths forward:

- We merge this targeted fix to resolve the immediate regression and create a separate ticket to address the larger refactoring project.
- We close this PR and revert the related commit (eb091731) if the team prefers to tackle the root cause directly. (Apologies to @mcasquer and @aliang123  for the potential inconvenience).

I would appreciate hearing everyone's thoughts on how we should proceed.

Thank you,

smitterl replied:
> `is_installed/check_installed` already in `package_install` logic, I don't think the `self.dependency_packages` part changes is needed. not sure what's the mean of "**also apply this early is_installed check to the dependencies**" 

What I meant to say: 

The changes addresses splitting installation check and installation but only for the stress package but there is also code 

```python
    def install(self):
        """
        To download, abstract, build and install the stress tool
        """
        # Install the dependencies before the tool gets installed
        if self.dependency_packages:
            if not utils_package.package_install(
                self.dependency_packages, session=self.session
            ):
                raise exceptions.TestError(
                    "Installing dependency packages for" " %s failed" % self.stress_type
                )
        if self.stress_package:
```

So, for consistency, this should also be split for `self.dependency_packages`. @YongxueHong addressed this.

PaulYuuu replied:
`dependency_packages` is only for stress build from source, when installing stress package from repo, dep list should be empty. `dependency_packages` is not the part of this PR, it handles correctly when build stress from srouce because deps are mandatory.

The current implementation reduce the performance. package_install support list, and the current `dependency_packages` is a list, this code introduce a logic to check them one by one, and use `package_install` to install one by one, which not make senses to me.
```mermaid
flowchart TD
    A1["Start"] --> B1{"dependency_packages exists?"}

    %% Existing code (first snippet)
    B1 -->|No| C1["Do nothing"]
    B1 -->|Yes| D1["Call package_install(all_packages)"]
    D1 --> E1["package_install internally checks is_installed"]
    E1 --> F1{"Installation successful?"}
    F1 -->|Yes| G1["Done"]
    F1 -->|No| H1["Raise TestError"]
    style D1 fill:#D5F5E3,stroke:#27AE60,stroke-width:1px
    style E1 fill:#D5F5E3,stroke:#27AE60,stroke-width:1px
    %% Greenish nodes = existing code

    %% PR modification (second snippet)
    B1 -->|Yes| I1["Create package manager"]
    I1 --> J1["Check is_installed manually"] 
    J1 --> K1["Filter dep_pkgs = not installed"]
    K1 --> L1{"dep_pkgs not empty?"}
    L1 -->|No| G1
    L1 -->|Yes| M1["Call package_install(dep_pkgs)"]
    M1 --> N1["package_install internally checks is_installed again (redundant!)"]
    N1 --> O1{"Installation successful?"}
    O1 -->|Yes| G1
    O1 -->|No| H1
    style I1 fill:#FCF3CF,stroke:#F1C40F,stroke-width:1px
    style J1 fill:#FAD7A0,stroke:#E67E22,stroke-width:1px
    style N1 fill:#F5B7B1,stroke:#C0392B,stroke-width:1px
    %% Yellow/Redish nodes = PR changes and redundant checks

```

YongxueHong replied:
> `dependency_packages` is only for stress build from source, when installing stress package from repo, dep list should be empty. `dependency_packages` is not the part of this PR, it handles correctly when build stress from srouce because deps are mandatory.

Hi @PaulYuuu
I do not have any idea about whether the `dependency_packages` is related to the `stress package` and whether both have an exclusive relationship.  I just keep the original design, not to break it.

Hi @smitterl
Could you help confirm what the relationship is between the `dependency_packages` and the `stress package`? Since I am not familiar with the background of it. Thanks.
CC @dzhengfy 

> 
> The current implementation reduce the performance. package_install support list, and the current `dependency_packages` is a list, this code introduce a logic to check them one by one, and use `package_install` to install one by one, which not make senses to me.

Hi @PaulYuuu 
Your consideration makes sense for me as well; however, I think it is another topic since the original implementation is to install the packages one by one:
`RemotePackageMgr`
https://github.com/avocado-framework/avocado-vt/blob/89ef3524dc4250d2f15811a321037f98b6b66c95/virttest/utils_package.py#L90-L121
`LocalPackageMgr`
https://github.com/avocado-framework/avocado-vt/blob/89ef3524dc4250d2f15811a321037f98b6b66c95/virttest/utils_package.py#L165-L182

We could improve the implementation of the installation in the future.
> 
> ```mermaid
> flowchart TD
>     A1["Start"] --> B1{"dependency_packages exists?"}
> 
>     %% Existing code (first snippet)
>     B1 -->|No| C1["Do nothing"]
>     B1 -->|Yes| D1["Call package_install(all_packages)"]
>     D1 --> E1["package_install internally checks is_installed"]
>     E1 --> F1{"Installation successful?"}
>     F1 -->|Yes| G1["Done"]
>     F1 -->|No| H1["Raise TestError"]
>     style D1 fill:#D5F5E3,stroke:#27AE60,stroke-width:1px
>     style E1 fill:#D5F5E3,stroke:#27AE60,stroke-width:1px
>     %% Greenish nodes = existing code
> 
>     %% PR modification (second snippet)
>     B1 -->|Yes| I1["Create package manager"]
>     I1 --> J1["Check is_installed manually"] 
>     J1 --> K1["Filter dep_pkgs = not installed"]
>     K1 --> L1{"dep_pkgs not empty?"}
>     L1 -->|No| G1
>     L1 -->|Yes| M1["Call package_install(dep_pkgs)"]
>     M1 --> N1["package_install internally checks is_installed again (redundant!)"]
>     N1 --> O1{"Installation successful?"}
>     O1 -->|Yes| G1
>     O1 -->|No| H1
>     style I1 fill:#FCF3CF,stroke:#F1C40F,stroke-width:1px
>     style J1 fill:#FAD7A0,stroke:#E67E22,stroke-width:1px
>     style N1 fill:#F5B7B1,stroke:#C0392B,stroke-width:1px
>     %% Yellow/Redish nodes = PR changes and redundant checks
> 
> ```

smitterl replied:
@YongxueHong IMO it doesn't matter if there's a relation of `dependency_packages` and the `stress package`. Your goal in the PR has been to avoid unnessary installation command run (your PR has title "
utils_test: optimize dependency and LocalPackageMgr"). For consistency shouldn't just do it for the stress package but also for the dependency_packages, s. https://github.com/avocado-framework/avocado-vt/pull/4208#discussion_r2330456168

YongxueHong replied:
> @YongxueHong IMO it doesn't matter if there's a relation of `dependency_packages` and the `stress package`. Your goal in the PR has been to avoid unnessary installation command run (your PR has title " utils_test: optimize dependency and LocalPackageMgr"). For consistency shouldn't just do it for the stress package but also for the dependency_packages, s. [#4208 (comment)](https://github.com/avocado-framework/avocado-vt/pull/4208#discussion_r2330456168)

Hi @PaulYuuu FYI

smitterl replied:
I'm sorry I don't understand this. Isn't the `is_installed` the decoupling? The `package_install` tries to install whereas the `is_installed` only checks if the package is already installed. I don't think this is a waste of time but I might be missing something. As discussed offline I think we should have a quick meeting to find an agreement. The original patch https://github.com/avocado-framework/avocado-vt/pull/4149 introduced a critical regression in tp-libvirt regression tests that use the stress package; this happened already a couple of months back. IMO we need to get this merged ASAP or revert https://github.com/avocado-framework/avocado-vt/pull/4149

YongxueHong commented:
> @YongxueHong, the second commit can be removed as those typos are fixed already.

Hi @PaulYuuu Thank you for your kind reminder.

dzhengfy reviewed (approved):
LGTM

mcasquer commented:
Stress affected test cases passed with latest PR code
